data/blog/perfectly-fitting-text-to-container-in-react.md

```diff
@@ -0,0 +1,295 @@
+
+This performance is acceptable to me. The resize happens fast enough to be imperceptible (faster than 100ms) is all cases. Moreover, since the data for the widgets loads async via HTTP, the UI has to watch for a `fetch()` call before the resize anyway, and compared to the `fetch()` duration, this resize duration is meaningless. Lastly, the `fetch()` calls all finish at different times, which amortizes the cost of the resize since they don't have to happen for all the elements at once.
+
+In the case of a page resize, they _do_ have to run all at once, but in my opinion that experience is still acceptable. Waiting a few hundred milliseconds on page resize is normal, and there are other more expensive operations running during a resize anyway.
+
+Success! I can ship, and reduce the max iteration count to 20, just to be generous.
+
+Oh wait, I mentioned there were bugs. It's true! I found some non-converging conditions that I caused with faulty logic. Easy enough.
+
+## Components and Their Props
+
+A small note on component API design. The first iteration of this component had this kind of setup:
+
+```jsx
+interface Props {
+  minFontSize: number;
+  maxFontSize: number;
+}
+
+function AutoSizeText(props: Props) {
+  const [fontSize, setFontSize] = useState < number > (maxFontSize - minFontSize) / 2
+
+  // More code
+}
+```
+
+I required a manual minimum and maximum for two reasons:
+
+1. Fear. I had some reservations about running this component without size restrictions. What if the algorithm fails, and the font size drops to 0, or otherwise explodes off the page? This fear is unfounded, I think. Why would such a thing happen? If it happens, I have telemetry that will catch it. If I see it in the wild, I should fix the condition that caused it.
+2. Simplicity. Asking for a manual size restriction makes the code easier to manage because I don't have to calculate the font size bounds myself. This is valid reason, but I don't think it's as important as DX.
+
+I ended up throwing those prop requirements. Requiring manual bounds had too many downsides, and it's a bad API.
+
+First of all, how would developers decide these bounds? Why wouldn't they always set the minimum as `0` and the maximum as some-impractically-large-number? Second of all, it creates unpleasant limits. The problem I ran into immediately is that I set a maximum bound too low, and was then confused why my auto-sized numbers weren't big enough. Third of all, this means there are _two_ APIs. The first API is through CSS, by creating a positioned parent element in HTML. The second API is through React's props. That's no good. I liked the idea of CSS as the main API, so I didn't want to _also_ have a second API to fiddle with.
+
+I ended up setting `minFontSize` to 0, and `maxFontSize` to the height of the parent component. Easy! The API is to set `position: relative;` (or any other position) on the parent, and `AutoSizedText` naturally fills the parent completely.
```

Comment:
❤️ 

There is another thing here, which is that you want to hide complexity of the component resize operations from the user and only expose them to a component which does exactly the right amount of resizing. If they need to tweak it manually via props, then the `Auto` in AutoSizeText isn't really that auto anymore and could even be considered a failure of the component itself.

Reply:
True! Added